### PR TITLE
Normalized function names (lower-case case-insensitive names)

### PR DESCRIPTION
**Pull request by amosbird**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Backward Incompatible Change

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Now all case-insensitive function names will be lower-cased during query analysis. This is needed for projection query routing.

Detailed description / Documentation draft:

missing documentation
`replace`

Some functions seem better to be case sensitive
- roundbankers
- yearweek
- extractallgroups

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
I think it will be more visually pleasant if they will be not lowercased, but normalized to canonical name.
Can we extract canonical name from function object (getName()) or from factory?

**Comment by alexey-milovidov** [comment]
Example: `extractallgroups` does not look beautiful.

**Comment by alexey-milovidov** [comment]
BTW, why extractAllGroups or roundBankers are case insensitive? It is simply a mistake.
Only functions compatible with SQL standard or MySQL/Postgres/SQLite, etc should be case insensitive.
ClickHouse native functions should not be case insensitive.

**Comment by amosbird** [comment]
Why are there so many flaky check failures...

**Comment by amosbird** [comment]
It seems function name normalization will break old instances when doing distributed query

**Comment by alexey-milovidov** [comment]
@amosbird Also please run `utils/generate-ya-make/generate-ya-make`.

File: src/Interpreters/FunctionNameNormalizer.cpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+const String & getFunctionCanonicalNameIfAny(const String & name);
+const String & getAggregateFunctionCanonicalNameIfAny(const String & name);
+
+void FunctionNameNormalizer::visit(IAST * ast)
+{
+    if (!ast)
+        return;
+
+    if (auto * node_storage = ast->as<ASTStorage>())
```
Comment (alexey-milovidov) [inline src/Interpreters/FunctionNameNormalizer.cpp:17]: Btw, why this code path is needed, and 

```
    for (auto & child : ast->children)
        visit(child.get());
```

is not enough?

File: src/Server/TCPHandler.cpp [hunk 2]
```diff
@@ -1133,6 +1133,12 @@ void TCPHandler::receiveQuery()
     }
     query_context->applySettingsChanges(settings_changes);
 
+    /// Disable function name normalization it's a secondary query.
```
Comment (alexey-milovidov) [inline src/Server/TCPHandler.cpp:1136]: Some explanation needed, e.g. "because they were already normalized on initiator node".

**Review by alexey-milovidov — COMMENTED** [review]
LGTM. Only a few changes and we can merge...

File: src/Interpreters/FunctionNameNormalizer.cpp (hunk 1 above)
Comment (amosbird) [reply src/Interpreters/FunctionNameNormalizer.cpp:17]: It normalizes selected children (avoid normalizing engine, Log is both engine name and log function). I'll add a comment.

File: src/Server/TCPHandler.cpp (hunk 2 above)
Comment (amosbird) [reply src/Server/TCPHandler.cpp:1136]: It's needed for old initiator new shard compatibility. Comments will be refined.

Comment (alexey-milovidov) [reply src/Server/TCPHandler.cpp]: Haha... yes, a comment is needed :)

**Pull request merged**